cpp/src/parquet/geospatial/statistics.cc

```diff
@@ -0,0 +1,392 @@
+
+    if (!is_valid_) {
+      return;
+    }
+
+    if (is_wraparound_x()) {
+      throw ParquetException("Wraparound X is not suppored by GeoStatistics::Update()");
+    }
+
+    ::arrow::Status status = ::arrow::internal::VisitSetBitRuns(
+        valid_bits, valid_bits_offset, num_spaced_values,
+        [&](int64_t position, int64_t length) {
+          for (int64_t i = 0; i < length; i++) {
+            ByteArray item = SafeLoad(values + i + position);
+            PARQUET_CATCH_NOT_OK(bounder_.MergeGeometry(
+                {reinterpret_cast<const char*>(item.ptr), item.len}));
+          }
+
+          return ::arrow::Status::OK();
+        });
+
+    if (!status.ok()) {
+      is_valid_ = false;
+    }
+  }
+
+  void Update(const ::arrow::Array& values) {
+    if (!is_valid_) {
+      return;
+    }
+
+    if (is_wraparound_x()) {
+      throw ParquetException("Wraparound X is not suppored by GeoStatistics::Update()");
+    }
+
+    // Note that ::arrow::Type::EXTENSION seems to be handled before this is called
+    switch (values.type_id()) {
+      case ::arrow::Type::BINARY:
+        UpdateArrayImpl<::arrow::BinaryArray>(values);
+        break;
+      case ::arrow::Type::LARGE_BINARY:
+        UpdateArrayImpl<::arrow::LargeBinaryArray>(values);
+        break;
+      // This does not currently handle run-end encoded, dictionary encodings, or views
+      default:
+        throw ParquetException("Unsupported Array type in GeoStatistics::Update(Array): ",
+                               values.type()->ToString());
+    }
+  }
+
+  void Reset() {
+    bounder_.Reset();
+    is_valid_ = true;
+  }
+
+  EncodedGeoStatistics Encode() const {
+    if (!is_valid_) {
+      return {};
+    }
+
+    const geospatial::BoundingBox::XYZM& mins = bounder_.Bounds().min;
+    const geospatial::BoundingBox::XYZM& maxes = bounder_.Bounds().max;
+
+    EncodedGeoStatistics out;
+
+    if (has_geometry_types_) {
+      out.geospatial_types = bounder_.GeometryTypes();
+    }
+
+    bool write_x = !bound_empty(0) && bound_valid(0);
+    bool write_y = !bound_empty(1) && bound_valid(1);
+    bool write_z = write_x && write_y && !bound_empty(2) && bound_valid(2);
+    bool write_m = write_x && write_y && !bound_empty(3) && bound_valid(3);
+
+    if (write_x && write_y) {
+      out.xmin = mins[0];
+      out.xmax = maxes[0];
+      out.ymin = mins[1];
+      out.ymax = maxes[1];
+      out.xy_bounds_present = true;
+    }
+
+    if (write_z) {
+      out.zmin = mins[2];
+      out.zmax = maxes[2];
+      out.z_bounds_present = true;
+    }
+
+    if (write_m) {
+      out.mmin = mins[3];
+      out.mmax = maxes[3];
+      out.m_bounds_present = true;
+    }
+
+    return out;
+  }
+
+  void Update(const EncodedGeoStatistics& encoded) {
+    if (!is_valid_) {
+      return;
+    }
+
+    // We can create GeoStatistics from a wraparound bounding box, but we can't
+    // update an existing one because the merge logic is not yet implemented.
```

Comment:
Would it be hard to implement? It's a bit weird and user-unfriendly to error out here.